### DISCUSSION TITLE
solver doesn't work after updating to Latest Pastas Version 0.16.0

### DISCUSSION THREAD

**Discussion by DataGeoranger** [pastas/pastas#274] [Q&A] [answered]
Hey,

I have generated a Pastas model that was running and produced decent fitting to the observed head. 
However, after I update to latest version, running the same code, the solve module completed failed and doesn't produce any matching (like a straight line). Then I give my code to my colleagues who hasn't updated to latest version, he was able to show good match. Could you please help? My model only has pumping stress, but I have 37 pumping wells.

Please see attached code and input dataset, it should run without any changing to the code. Thank you very much and look forward to hearing from you!

Best,
Jack

![image](https://user-images.githubusercontent.com/47303462/109408089-406c8680-7954-11eb-9bde-663e245c37e5.png)

[Coordinates_Distances.xlsx](https://github.com/pastas/pastas/files/6056264/Coordinates_Distances.xlsx)
[Flow_Rates.xlsx](https://github.com/pastas/pastas/files/6056265/Flow_Rates.xlsx)
`[139.xlsx](https://github.com/pastas/pastas/files/6056266/139.xlsx)`

```
import matplotlib.pyplot as plt
import pandas as pd
import os
import pastas as ps
from pathlib import Path
import numpy as np

ps.show_versions()

os.chdir(os.path.dirname(os.path.abspath(__file__)))

# import distance of EWs to MW
distances = pd.read_excel("Coordinates_Distances.xlsx", sheet_name = 'distances_2')
distances_139 = distances['139']
dist_list = distances_139.tolist()
type(dist_list)

# import MW observed head
gw = pd.read_excel('139.xlsx', parse_dates=['Date'], index_col = 'Date', squeeze=True)
gw.plot()

# import pumping rates
FR1 = pd.read_excel('Flow_Rates.xlsx', sheet_name='rates-cfd', 
                    parse_dates=['Date'], index_col = 'Date', squeeze=True)
FR1.plot(figsize = (25, 20))

EWs = ['4060',	'4070',	'4075',	'4140',	'4145',	'4200',	'4315',	'4425',	'4435',	'4540',	'4545',	'4550',	
'4555',	'4590',	'4595',	'4600',	'4605',	'4610',	'4615',	'4635',	'4640',	'4645',	'4650',	'4655',	'4665',	
'4680',	'4685',	'4690',	'4695',	'4730',	'4744',	'4747',	'4748',	'4754',	'4755',	'1098',	'2067',	'2069']

# create Pastas model
ml = ps.Model(gw, name="groundwater")

# adding stresses
stresses1 = []
for name in EWs:
    stress = ps.TimeSeries(FR1.loc[:, name], name=name, settings='well')
    print(name)
    print(FR1.loc[:, name])
    stresses1.append(stress)

# add to well model
w = ps.WellModel(stresses1, ps.HantushWellModel, "Wells", distances = dist_list, settings="well")
ml.add_stressmodel(w)

# Solve
ml.solve(warmup = 1, freq = "h", initial=0.01, solver=ps.LmfitSolve)

# Plot
ml.plots.decomposition()
```

**Comment by raoulcollenteur** [comment, marked as answer]
Hi Jack,

Strange, although there were some changes to the noise model that might explain this. Can you run `ps.show_versions()` and paste the package versions here? And a screenshot of the fit_report from the model with Pastas 0.15.0? 

Cheers,
Raoul




ml.solve(warmup = 1, freq = "h", solver=ps.LmfitSolve)

**Reply by DataGeoranger** [reply]
Hey Raoul,
Sorry, it took me a while to reply. It was complicated when downgrading Pastas version with new numpy version also released recently, it was a mess. Anyway, please see below the screenshot of package version and fit report using Pastas 0.16.0 and Pastas 0.14.0 (My colleague was running 0.14.0 and was fitting well). My colleague and I didn't change the code or dataset at all when running it, it was just running different version of Pastas. And we are getting complete different results. Thanks so much for looking into this!

Best,
Jack
Screenshot of the fit report from model with Pastas 0.16.0 version
![image](https://user-images.githubusercontent.com/47303462/109515446-1363c500-7a75-11eb-8785-8249553ed656.png)
![image](https://user-images.githubusercontent.com/47303462/109515487-1fe81d80-7a75-11eb-9b86-dcd94067eff3.png)
 
Screenshot of fit report from model with Pastas 0.14.0 version.
![image](https://user-images.githubusercontent.com/47303462/109518366-2d52d700-7a78-11eb-925c-421da80bb570.png)
![image](https://user-images.githubusercontent.com/47303462/109518059-ea90ff00-7a77-11eb-8ed3-8a6c73834713.png)

**Reply by dbrakenhoff** [reply]
Hi Jack,

I've also tried reproducing the different results between pastas versions, and I've come up with the following results:

- Pastas 0.14.0, freq="H", EVP=87% with noise=False (fit is no good with noise=True)
- Pastas 0.14.0, freq="D", EVP=87% with noise=True 
- Pastas (dev branch), freq="D", EVP=0%, regardless of noise setting

I'll try taking a look later whether I can figure out why the fit isn't achieving similar results in the most recent pastas version. My guess is it has something to do with initial parameter values in WellModel. Between pastas 0.14.0 and 0.16.0 the parameter definitions in HantushWellModel changed slightly so that might be influencing the result... 

By the way, the solve call in your example code is a bit odd, and seemingly different from the earlier result with pastas 0.14.0:

```python
ml.solve(warmup = 1, freq = "h", initial=0.01, solver=ps.LmfitSolve)
```

In the screenshot of the model with a decent fit result the warmup is equal to the default setting (3650 days) and the frequency is "D". The initial keyword argument should be a boolean value, that determines whether to recalculate (reset) the initial values (from the defaults) (initial=True), or use the current optimal values as a starting point for the optimization (`initial=False`).

**Reply by DataGeoranger** [reply]
Hi David,

Thank you for looking into this and please let me know when you had a chance to look into what causes changes in fitting between 0.14.0 and 0.16.0.

Good catch on the `ml.solve` parameter setting. I was trying different parameter to see if I can get a decent fit in 0.16.0, and none of the warmup, initial was sensitive. Because my dataset only spans 2 months, I was suspecting if a too long of warmup like 3650 days would cause the solver to fail. Thanks for the explanation of initial, I thought that was a real number for a parameter, but somehow Pastas proceeded when I set it to 0.01. Will notice that in the future.

I do have three follow up questions for the unit of the pumping well rate and gain numbers in the multiple well example [10_multiple_wells_nogainparameter.ipynb](https://github.com/pastas/pastas/blob/master/examples/notebooks/10_multiple_wells.ipynb). 

Questions 1: All the existing example was in meter, cubic meter per day. The unit for my project are in feet and cubic feet per day. Does Pastas only work with metric units or it doesn't matter as long as all my input are consistent in **L** and **T**? 

Question 2: My observation head is recorded every 10 mins and my pumping rate is in cubic feet per day. Do I need to prepare the pumping input dataset that has pumping rate unit in cubic feet per 10 min, so similar to my observation head, or Pastas will refine the time interval of pumping stress during solving?

Question 3: In the example notebook, the Pastas will calculate the Gain number in the unit of meter for the Hantush response function. But my understanding is that the gain is not drawdown, but an indicator that represents the level of impact from a pumping well to the observation well. Is the understanding correct? I am trying to understand the number of the gain that calculated in the example notebook.

Thank you very much again and look forward to your reply!

Best,
Jack

**Reply by dbrakenhoff** [reply]
Hi Jack,

> Questions 1: All the existing example was in meter, cubic meter per day. The unit for my project are in feet and cubic feet per day. Does Pastas only work with metric units or it doesn't matter as long as all my input are consistent in L and T?

Pastas does not care about the units, so you can work with feet and cubic feet per day. 

> Question 2: My observation head is recorded every 10 mins and my pumping rate is in cubic feet per day. Do I need to prepare the pumping input dataset that has pumping rate unit in cubic feet per 10 min, so similar to my observation head, or Pastas will refine the time interval of pumping stress during solving?

By default the pastas model frequency (or timestep) is set to "days". This means that if you provide 10-minute data for the observation head and call `ml.solve()`, Pastas will select one value per day, and use that value to calibrate the model. If you want to calibrate on all your data points you would set `ml.solve(freq="10T")`. This also means your stresses must have a value every 10 minutes. When solving using this frequency Pastas will automatically resample your extraction timeseries based on the settings for that timeseries. So it will create an extraction timeseries with values every 10 minutes. 

However, please note that calibrating on 10-minute data is probably not a great idea though, it is computationally expensive, and will likely take long time to run. Also most groundwater processes are relatively slow so there is no real benefit in calibrating the model on this level of detail (plus you might run into autocorrelation issues if you want to assess parameter uncertainty). Your resulting model will almost certainly be better if you stick to daily timesteps (unless there is some extremely fast process you really want to take into account). 

> Question 3: In the example notebook, the Pastas will calculate the Gain number in the unit of meter for the Hantush response function. But my understanding is that the gain is not drawdown, but an indicator that represents the level of impact from a pumping well to the observation well. Is the understanding correct? I am trying to understand the number of the gain that calculated in the example notebook.

The gain is the final increase (or decrease) that results from a unit increase or decrease in a stress that continues infinitely in time (e.g., pumping at a constant rate forever). If your extraction rate is given in cubic feet/day, and your observations are in feet, the gain represents the lowering of the head if you extract 1 cubic feet per day forever. The units of the gain would then be feet / (cubic feet per day). So it's not quite drawdown but could be seen as the steay-state drawdown for an extraction with magnitude 1. 

Hope this answers your questions?

**Reply by DataGeoranger** [reply]
@dbrakenhoff Thank you so much again for the detailed reply!. It helps a ton! 
Just follow up questions on questions 2 and 3: 

> Question 2: My observation head is recorded every 10 mins and my pumping rate is in cubic feet per day. Do I need to prepare the pumping input dataset that has pumping rate unit in cubic feet per 10 min, so similar to my observation head, or Pastas will refine the time interval of pumping stress during solving?

> By default the pastas model frequency (or timestep) is set to "days". This means that if you provide 10-minute data for the observation head and call ml.solve(), Pastas will select one value per day, and use that value to calibrate the model. If you want to calibrate on all your data points you would set ml.solve(freq="10T"). This also means your stresses must have a value every 10 minutes. When solving using this frequency Pastas will automatically resample your extraction timeseries based on the settings for that timeseries. So it will create an extraction timeseries with values every 10 minutes.

> However, please note that calibrating on 10-minute data is probably not a great idea though, it is computationally expensive, and will likely take long time to run. Also most groundwater processes are relatively slow so there is no real benefit in calibrating the model on this level of detail (plus you might run into autocorrelation issues if you want to assess parameter uncertainty). Your resulting model will almost certainly be better if you stick to daily timesteps (unless there is some extremely fast process you really want to take into account).

In my case, the transducer study lasts 11 days and monitoring wells are just nearby the pumping well, I was wondering if I need a more refined time step to be able to capture the fluctuation of water levels that happens within a couple hours. Would you suggest a finer time frequency for a shorter period of water levels like a pumping test etc.? 

Just to confirm I understand correctly about the resampling my extraction timeseries. So if my input unit for extraction is cubic feet per day, Pastas will resample so that there is an input series of extraction with unit of cubic feet per 10 min? If so, is there a function in Pastas that allow me to export the simulated extraction rate with unit of cubic feet per 10 min that Pastas resampled based on frequency of 10 min? It will be helpful information to QA because I wasn't sure what was simulated, so I even prepared an extraction input with cubic feet per 10 mins.

> Question 3: In the example notebook, the Pastas will calculate the Gain number in the unit of meter for the Hantush response function. But my understanding is that the gain is not drawdown, but an indicator that represents the level of impact from a pumping well to the observation well. Is the understanding correct? I am trying to understand the number of the gain that calculated in the example notebook.

> The gain is the final increase (or decrease) that results from a unit increase or decrease in a stress that continues infinitely in time (e.g., pumping at a constant rate forever). If your extraction rate is given in cubic feet/day, and your observations are in feet, the gain represents the lowering of the head if you extract 1 cubic feet per day forever. The units of the gain would then be feet / (cubic feet per day). So it's not quite drawdown but could be seen as the steady-state drawdown for an extraction with magnitude 1.

The observed drawdown in my monitoring well is about 0.5 feet. So if I sum the gain from each extraction well, would I expect summation of the gain equal to the observed drawdown in my monitoring wells? If I set my model frequency to 10 mins, would this gain have an unit of  feet/(1 cubic feet per 10 min)?

Thanks again for your time and help!

Best,
Jack

**Reply by dbrakenhoff** [reply]
> In my case, the transducer study lasts 11 days and monitoring wells are just nearby the pumping well, I was wondering if I need a more refined time step to be able to capture the fluctuation of water levels that happens within a couple hours. Would you suggest a finer time frequency for a shorter period of water levels like a pumping test etc.?

For a pumping test a smaller timestep might indeed be relevant. In that case you want the model to not only get the steady-state drawdown right, but also the shape of the drawdown curve. It might also be smart to manually create an observations heads series for calibration in which the initial drawdown right after turning on the pumping well is adequately represented in the data (i.e. ensure you have sufficient heads observations in that initial drawdown period).

> Just to confirm I understand correctly about the resampling my extraction timeseries. So if my input unit for extraction is cubic feet per day, Pastas will resample so that there is an input series of extraction with unit of cubic feet per 10 min? If so, is there a function in Pastas that allow me to export the simulated extraction rate with unit of cubic feet per 10 min that Pastas resampled based on frequency of 10 min? It will be helpful information to QA because I wasn't sure what was simulated, so I even prepared an extraction input with cubic feet per 10 mins.

Pastas resamples based on the timeseries settings. There are different options but you're probably using "well", which means Pastas will resample using backfill. This means the daily value is just copied backwards to fill the 10-minute steps. You can take a look at how pastas deals with this with the following code:

```python
ts = ps.Timeseries(well, name="well", settings="well")
ts.update_series(freq="10T", sample_up="bfill")
ts.original_series  # the original series
ts.series  # contains the updated series
```
We often recommend to pre-process the timeseries yourself so you know exactly what you put in, that way it's also easier to understand the output. Also the following notebook might help clear things up a bit: https://pastas.readthedocs.io/en/dev/concepts/handling_timeseries.ipynb.html

> The observed drawdown in my monitoring well is about 0.5 feet. So if I sum the gain from each extraction well, would I expect summation of the gain equal to the observed drawdown in my monitoring wells?

If you sum the gain for each well, you would get the drawdown if each well was pumping at 1 cubic foot per day. So for what you're saying to be true, you'll have to multiply the gain by the mean extraction rate per well and then add them all together.

If you change the model frequency to 10 minutes, your extraction timeseries will still have the same units (cubic feet per day) but will have a value every 10 minutes. The gain will still be in feet / (1 cubic foot per day). We chose to do it this way because it makes it easier to compare models with different frequencies when the units stay the same.

**Reply by DataGeoranger** [reply]
Thank you again for the wonderful insights @dbrakenhoff 
> Pastas resamples based on the timeseries settings. There are different options but you're probably using "well", which means Pastas will resample using backfill. This means the daily value is just copied backwards to fill the 10-minute steps. You can take a look at how pastas deals with this with the following code:

I experiment various frequency setting by varying my Pastas model frequency to "10T", "H", "D" 
` ml.settings["freq"] = "1H"`
without changing the input of my extraction rate - which is daily extraction rate in cubic feet per day. If the backfill works correctly, I am guessing I will be getting the similar matching between different frequencies of 10 mins, 1 hour and 1 day. However, the results shows quite different matching results between "10T", "H", "D", where "D" frequency gives me the best match, and the match from "10T" and "H" are quite off. Please see screenshot below. I wonder what could cause such difference?  
##Matching curve of 10T
![image](https://user-images.githubusercontent.com/47303462/110536574-9e853080-80ef-11eb-9515-32c36f5e160d.png)
##Matching curve of 1H
![image](https://user-images.githubusercontent.com/47303462/110536650-b2309700-80ef-11eb-8bc2-a0727ce118d7.png)
## Matching curve of D
![image](https://user-images.githubusercontent.com/47303462/110536704-c5436700-80ef-11eb-8e89-ef07aa6a7344.png)

> ts = ps.Timeseries(well, name="well", settings="well")
> ts.update_series(freq="10T", sample_up="bfill")
> ts.original_series  # the original series
> ts.series  # contains the updated series

I couldn't use Timeseries method to see how Pastas converting the rates internally. Because I am using version 0.14.0. Recall that I had to downgrade from 0.16.0 to 0.14.0 because 0.16.0 doesn't produce the same results as 0.14.0 after the updates as we discussed earlier.

**Reply by dbrakenhoff** [reply]
Hi Jack,

I'm able to get fairly close to the result you're getting with 0.14.0 with the latest version of Pastas. By playing around with the initial values of the parameters I was able to get a decent fit with freq="H".

In the table below are the initial and optimal parameter values as calculated by pastas version 0.14.0 and version 0.17b.0 (latest develop version). For parameter `Wells_b` marked in  _**bold and italic**_ I used the initial value from pastas 0.14.0 instead of the default value calculated by pastas 0.17b.0. When I modified that, I was able to get a reasonable fit (EVP = 80%). Note that there are still pretty big differences between the optimal parameters, but you' ll have to determine which model you think is better... 

|             |   initial_017 |   optimal_017 |   initial_014 |   optimal_014 |
|:------------|--------------:|--------------:|--------------:|--------------:|
| Wells_A     |  -4.24633e-06 |  -6.90853e-06 |  -4.24633e-06 |  -3.39851e-06 |
| Wells_a     | 100           |  12.5714      | 100           |  39.6767      |
| Wells_b     |   **_1.875340e-08_** |   5.08144e-09 |   **_1.08438e-09_** |   1.25036e-09 |
| constant_d  |   2.869       |   7.60565     |   2.869       |  11.8184      |
| noise_alpha |   1           |   8.64836     |   1           |   1.88129     |

The old parameters were `A`, `lab` and `cS`, the new ones are `A`, `a`, `b`. This is how they relate to one another for HantushWellModel:

- `A = A`
- `a = cS`
- `b = 1 / (4 * lab**2)`

Hope that helps you on your way. I'd definitely retry the analysis you did above using the latest version of pastas. I'm fairly confident v0.16.0 will work similar to the develop version (v0.17b.0) that I used. 

Cheers,

Davíd

**Reply by DataGeoranger** [reply]
Hi David,

You @dbrakenhoff are absolutely awesome. Thanks for the tips and reply. I upgraded to 0.16.0 and tried to reproduce your results in the reply. How did you set the value for the initial parameters? I went to the `rfunc.HantushWellModel` and only found the method for `get_init_parameters()`. 
I also try to retrieve the initial parameter and optimal parameters, and didn't work. 
`ps.rfunc.HantushWellModel.get_init_parameters(ml,"HantushWellModel")`
![image](https://user-images.githubusercontent.com/47303462/111242166-fb438800-85d4-11eb-8069-a414495eab90.png)

How did you actually obtain those parameter values using Pastas? Thank you for your patience with me, I feel I am really close to reproduce the results with 0.16.0 and be able to adjust the parameters moving forward.

I was also able to retrieve the extraction dataset from Pastas using the code below. I found there are a little bit question mark with backfill method in Pastas. Screenshot 1 is my daily rate and Screenshot 2 is the update series (every 10 mins) this case. You can see the backfill method didn't use the extraction rate at the beginning (10-01-2020) and extraction rate at the end (12-31-2020). Like you mentioned earlier. It is probably a good idea for the user to provide own input dataset. 

Best,
Jack
```
ts = ps.TimeSeries(FR1.loc[:, '4060'], name="well", settings="well")
ts.update_series(freq="10T", sample_up="bfill")
ts.series_original  # the original series
ts.series  # contains the updated series
```
### Screenshot 1
![image](https://user-images.githubusercontent.com/47303462/111241820-4741fd00-85d4-11eb-914e-4a5d7d43585e.png)
### Screenshot 2
![image](https://user-images.githubusercontent.com/47303462/111241850-51fc9200-85d4-11eb-8248-1adac97da0d7.png)

**Reply by dbrakenhoff** [reply]
This is the code I used: use `ml.set_parameter` to change a parameters initial value, bounds etc. You can check `ml.parameters` to see which initial values were used, and see all the other information about the parameters.

```python
ml = ps.Model(gw_sample, name="groundwater")
ml.settings["freq"] = "H"

# adding stresses
stresses1 = []
for name in EWs:
    stress = ps.TimeSeries(FR1.loc[:, name], name=name, settings='well')
    stresses1.append(stress)

# add to well model
w = ps.WellModel(stresses1.copy(), ps.HantushWellModel, "Wells",
                 distances=np.copy(dist_list), settings="well", up=False)
# add stressmodel
ml.add_stressmodel(w)

# set optimal value as calculated by pastas 0.14.0 as initial
ml.set_parameter("Wells_b", initial=1 / (4 * 1.41400808e+04**2))

# view initial parameters
print(ml.parameters)

ml.solve(freq="H", solver=ps.LmFitSolve)

# view parameters with optimal values this time
print(ml.parameters)
```

As for the timeseries backfill, pastas assumes the measurement time is at the end of the period being measured. Therefore the values of 2020-10-02 00:00 are assumed to represent the measurement over the previous day = 2020-10-01. That's why you're seeing that result. If you want it differently its up to you as the user to modify the timeseries.

**Reply by DataGeoranger** [reply]
@dbrakenhoff Thank you David! I was able to reproduce your results with frequency = H and get EVP = 79.38% with 0.16.0. I was able to improve the EVP to 83.42% with frequency = 10 mins. 

Thanks for the explanation of backfill. If my daily pumping rate is 100 cubic feet per day, it seems that the backfill method generates a 10 min time series dataset, with the extraction rates equal to the daily values. 100cubic feet per day for every 10 mins. 

However, from a volume perspective, the total volume was extracted during a day equal 100 cubic feet per day multiply 1 day, to be 100 cubic feet. To calculate the volume is extracted per 10 min, it also make sense to divide the total volume (100 cubic feet) by the 144 (how many 10 min in a day) to calculate the how much volume is extracted every 10 mins. I am not sure what is the best way to prepare pumping rate. 

Best,
Jack

**Reply by dbrakenhoff** [reply]
Good to hear you got it working!

As for your thoughts on the pumping rate, both approaches are fine. The reason we don't divide the rate (or volume) is so that the parameters of the models for different frequencies have the same units and are easier to compare. But it would be fine if you wanted to divide the pumped volume over the 10 minute intervals, but this will mean you'll have to convert the parameters yourself if you want to compare them.

**Reply by DataGeoranger** [reply]
That makes sense now and would actually explain some differences in results when I used my own input. I'll look into that!
Hey David, I just want to thank you again for your great support and discussion. I really appreciate it and am exciting about the tool!

Best Regards,
Jack
